packages/mobile/src/screens/search-screen-v2/screens/FilterBpmScreen.tsx

```diff
@@ -0,0 +1,338 @@
+      } else if (maxError) {
+        setMaxError(null)
+      }
+    } else if (maxError) {
+      setMaxError(null)
+    }
+
+    if (hasError) return
+
+    // Value Update
+    let value = ''
+
+    if (minBpm || maxBpm) {
+      value = `${minBpm || MIN_BPM}-${maxBpm || MAX_BPM}`
+    }
+
+    onChange(value)
+  }, [maxBpm, minBpm, minError, maxError, onChange])
+
+  return (
+    <>
+      <Flex>
+        <ListSelectionItems
+          data={rangeOptions}
+          value={bpmRange}
+          onChange={setBpmRange}
+        />
+      </Flex>
+      <Flex
+        flex={1}
+        direction='row'
+        p='l'
+        gap='xs'
+        w='100%'
+        alignItems='flex-start'
+      >
+        {bpmRange === 'custom' ? (
+          <>
+            <TextInput
+              style={{ flex: 1, alignSelf: 'flex-start' }}
+              value={minBpm}
+              label={messages.minBpm}
+              keyboardType='numeric'
+              error={!!minError}
+              helperText={minError}
+              aria-errormessage={minError ?? undefined}
+              placeholder={messages.minBpm}
+              onChange={(e) => setMinBpm(e.nativeEvent.text)}
+            />
+            <Text style={{ alignSelf: 'flex-start', paddingVertical: 20 }}>
+              -
+            </Text>
+            <TextInput
+              style={{ flex: 1, alignSelf: 'flex-start' }}
+              value={maxBpm}
+              label={messages.maxBpm}
+              keyboardType='numeric'
+              error={!!maxError}
+              helperText={maxError}
+              aria-errormessage={maxError ?? undefined}
+              placeholder={messages.maxBpm}
+              onChange={(e) => setMaxBpm(e.nativeEvent.text)}
+            />
+          </>
+        ) : null}
+      </Flex>
+    </>
+  )
+}
+
+const BpmTargetView = ({ value, setValue }: ViewProps) => {
+  const { color } = useTheme()
+  const [bpmTarget, setBpmTarget] = useState('')
+  const [bpmTargetType, setBpmTargetType] = useState<BpmTargetType>('exact')
+  const [error, setError] = useState<string | null>(null)
+  // NOTE: Memo to avoid the constantly changing function instance from triggering the effect
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const onChange = useMemo(() => setValue, [])
+
+  useEffect(() => {
+    if (!value && bpmTarget) {
+      setBpmTarget('')
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [value])
+
+  useEffect(() => {
+    // Validation
+    const val = Number(bpmTarget)
+    let hasError = false
+
+    if (bpmTarget) {
+      if (val < MIN_BPM) {
+        setError(messages.tooLowError)
+        hasError = true
+      } else if (val > MAX_BPM) {
+        setError(messages.tooHighError)
+        hasError = true
+      } else if (error) {
+        setError(null)
+      }
+    } else if (error) {
+      setError(null)
+    }
+
+    if (hasError) return
+
+    // Value Update
+    let value = ''
+
+    if (bpmTarget) {
```

Comment:
I kinda think if we can use HarmonyTextField, then most of the validation can be covered with a zod schema, and value transform can be done with `transformValueOnChange` prop.